kubernetes/deployments_kubeclient.go

```diff
@@ -0,0 +1,1339 @@
+	} else if result == nil {
+		return nil, nil
+	}
+	// Find the current deployment for the DC we just found. This should correspond to the deployment
+	// shown in the OpenShift web console's overview page
+	rcs, err := kc.getReplicationControllers(namespace, result.dcUID)
+	if err != nil {
+		return nil, errs.WithStack(err)
+	} else if len(rcs) == 0 {
+		return result, nil
+	}
+
+	// Find newest RC created by this DC, which is also considered visible according to the
+	// OpenShift web console's criteria:
+	// https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/controllers/overview.js#L658
+	const deploymentPhaseAnnotation string = "openshift.io/deployment.phase"
+	var newest *v1.ReplicationController
+	for idx := range rcs {
+		rc := &rcs[idx]
+		if newest == nil || newest.CreationTimestamp.Before(rc.CreationTimestamp) {
+			newest = rc
+		}
+	}
+	if newest != nil {
+		result.current = newest
+	}
+	return result, nil
+}
+
+func (kc *kubeClient) getReplicationControllers(namespace string, dcUID types.UID) ([]v1.ReplicationController, error) {
+	rcs, err := kc.ReplicationControllers(namespace).List(metaV1.ListOptions{})
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+
+	// Current Kubernetes concept used to represent OpenShift Deployments
+	rcsForDc := []v1.ReplicationController{}
+	for _, rc := range rcs.Items {
+
+		// Use OwnerReferences to map RC to DC that created it
+		match := false
+		for _, ref := range rc.OwnerReferences {
+			if ref.UID == dcUID && ref.Controller != nil && *ref.Controller {
+				match = true
+			}
+		}
+		if match {
+			rcsForDc = append(rcsForDc, rc)
+		}
+	}
+
+	return rcsForDc, nil
+}
+
+func (kc *kubeClient) getResourceQuota(namespace string) (*app.EnvStats, error) {
+	const computeResources string = "compute-resources"
+	quota, err := kc.ResourceQuotas(namespace).Get(computeResources, metaV1.GetOptions{})
+	if err != nil {
+		return nil, errs.WithStack(err)
+	} else if quota == nil {
+		return nil, errs.Errorf("no resource quota with name: %s", computeResources)
+	}
+
+	// Convert quantities to floating point, as this should provide enough
+	// precision in practice
+	cpuLimit, err := quantityToFloat64(quota.Status.Hard[v1.ResourceLimitsCPU])
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+	cpuUsed, err := quantityToFloat64(quota.Status.Used[v1.ResourceLimitsCPU])
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+
+	cpuStats := &app.EnvStatCores{
+		Quota: &cpuLimit,
+		Used:  &cpuUsed,
+	}
+
+	memLimit, err := quantityToFloat64(quota.Status.Hard[v1.ResourceLimitsMemory])
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+
+	memUsed, err := quantityToFloat64(quota.Status.Used[v1.ResourceLimitsMemory])
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+
+	memUnits := "bytes"
+	memStats := &app.EnvStatMemory{
+		Quota: &memLimit,
+		Used:  &memUsed,
+		Units: &memUnits,
+	}
+
+	result := &app.EnvStats{
+		Cpucores: cpuStats,
+		Memory:   memStats,
```

Comment:
you can also write it as:
````
result := &app.EnvStats{
 		Cpucores: cpuStats,
 		Memory:  &app.EnvStatMemory{
   			Quota: &memLimit,
 			Used:  &memUsed,
 			Units: &memUnits,
 		},
}
````
(if you like)

Reply:
Yes @xcoulon this is recommended!

Reply:
Changed.

Reply:
ok